crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -11,48 +11,133 @@
 //! arguments must match _at least one_ overload.
 
 use super::{definition_expression_type, DynamicType, Type};
+use crate::semantic_index::definition::Definition;
+use crate::symbol::Boundness;
+use crate::types::todo_type;
 use crate::Db;
-use crate::{semantic_index::definition::Definition, types::todo_type};
 use ruff_python_ast::{self as ast, name::Name};
 
+/// The signature of a possible union of callables.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, salsa::Update)]
+pub(crate) struct Signatures<'db> {
+    /// The type that is (hopefully) callable.
+    pub(crate) ty: Type<'db>,
+    /// For an object that's callable via a `__call__` method, the type of that method. For an
+    /// object that is directly callable, the type of the object.
```

Comment:
hmm, what's meant by "directly callable" here? _All_ callable objects in Python have a `__call__` method, even functions; this is core to the definition of what it means for an object to be callable in the language:

```py
>>> def foo():
...     print('hi')
...     
>>> foo.__call__()
hi
```

For _some_ types, like function-literal types, we don't attempt to lookup the `__call__` method in red-knot in order to infer the signature of the callable. But that doesn't mean that the `__call__` method doesn't exist!

Reply:
Ah that's a good clarification!  The idea I'm trying to convey is whether we use type of the object or the type of the `__call__` method in certain error messages.  Is there existing terminology that we can use here?